Add workqueue support

cizezsy opened the pull request:
Add workqueue support, see #523. Feel free to close this PR, if you think #525 is better :P

Will add more unit tests later.

yue9944882 commented:
/cc @adohe

he's getting back on wednesday, let's hold until some input from him. by skimming the sketch, i find the rate-limiter enhancement compelling. do you consider port [the leaky-bucket-based rate-limiter](https://github.com/kubernetes/kubernetes/blob/b69cc78e5807e30ef9406e16e07fc87bd109cdcf/staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go#L48-L63) to the tree?

cizezsy commented:
@yue9944882 Added it :P I will add more unit tests.

Also, the BucketRateLimiter is based on token bucket according to the documentation of rate.Limiter
>  It implements a "token bucket" of size b, initially full and refilled at rate r tokens per second.

brendandburns commented on extended/src/main/java/io/kubernetes/client/extended/workqueue/DelayingWorkQueue.java:
```diff
@@ -0,0 +1,92 @@
+package io.kubernetes.client.extended.workqueue;
```
Why not just use `ScheduledThredPoolExecutor`?

https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ScheduledThreadPoolExecutor.html

I think it's better to use built-ins unless absolutely necessary...

yue9944882 replied:
let's use the work-queue implementation from another pull #524

brendandburns commented on extended/src/main/java/io/kubernetes/client/extended/workqueue/WorkQueue.java:
```diff
@@ -0,0 +1,122 @@
+  // queue defines the order in which we will work on items. Every
+  // element of queue should be in the dirty set and not in the
+  // processing set.
+  private LinkedList<T> queue;
```
Let's use a `BlockingQueue` here

(https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/BlockingQueue.html)

To limit the potential memory size.

yue9944882 replied:
let's use the work-queue implementation from another pull #524

brendandburns commented on extended/src/main/java/io/kubernetes/client/extended/workqueue/ratelimiter/RateLimiter.java:
```diff
@@ -0,0 +1,23 @@
+
+import java.time.Duration;
+
+public interface RateLimiter<T> {
```
Why not use RateLimiter from Guava?

https://github.com/google/guava/blob/master/guava/src/com/google/common/util/concurrent/RateLimiter.java

yue9944882 replied:
the guava rate-limiter's API is different from the one we want for work-queue here i suppose. guava's rate-limiter is basically expecting an invocation to be blocked until permitted, while for work-queue, the so-called `RateLimiter` here is more like a backoff-provider, it will return how much time an inbound item need to be waiting for next round of processing.. probably we should consider rename this to sth like `BackOffGetter`, or `BackOffFactory`? @brendandburns @cizezsy wdyt?

cizezsy replied:
Yeah, I have looked at Guava RateLimiter before, but it does not satisfy our requirement. 

Regard the name, maybe we should keep the name same with client-go's so that others can quickly know how to use client-java?

brendandburns commented:
I put in some suggestions about existing libraries/classes you could use instead of re-implementing things...

yue9944882 commented:
>  you could use instead of re-implementing things...

+1 on we should avoid re-implementing wheels as much as possible

yue9944882 commented:
@cizezsy another work-queue has got in, how do you think about rebase and replace the rate-limiter onto that?

cizezsy commented:
@yue9944882 ok, works for me :P

cizezsy commented:
@yue9944882 rebased and added some unit test :)

yue9944882 commented on extended/src/main/java/io/kubernetes/client/extended/workqueue/ratelimiter/ItemExponentialFailureRateLimiter.java:
```diff
@@ -0,0 +1,51 @@
+  }
+
+  @Override
+  public synchronized Duration when(T item) {
```
```suggestion
  @Override
  public synchronized Duration when(T item) {
    int exp = failures.getOrDefault(item, 0);
    failures.put(item, exp + 1);
    long d = maxDelay.toMillis() >> exp;
    return d > baseDelay.toMillis() ? baseDelay.multipliedBy(1 << exp) : maxDelay;
  }
```

how do you think about this approach?

adohe replied:
IMHO, Java has many built-in thread-safe data structure, try to avoid using synchronized method here, `java.util.concurrent.ConcurrentHashMap` should be a good fit for this case, and is much more high efficient.

cizezsy replied:
I think `ConcurrentHashMap` is not thread-safe when we are doing multiple operations...

brendandburns replied:
You can use a combination of `ConcurrentHashMap` and `AtomicInteger` here.

cizezsy replied:
OK, replaced with AtomicLongMap now.

yue9944882 commented on extended/src/main/java/io/kubernetes/client/extended/workqueue/ratelimiter/BucketRateLimiter.java:
```diff
@@ -0,0 +1,56 @@
+import java.time.Duration;
+
+/** A light-weight token bucket implementation for RateLimiter. */
+public class BucketRateLimiter<T> implements RateLimiter<T> {
```
thanks, the new bucketing library looks compelling. i tried simplify implemtation of `when` a bit learning from https://github.com/vladimir-bukhtoyarov/bucket4j/issues/96, wdyt?
 
```suggestion
public class BucketRateLimiter<T> implements RateLimiter<T> {

  private Bucket bucket;
  private long tokensInQueue;
  private long tokensGeneratedInPeriod;
  private Duration period;

  /**
   * @param capacity                Capacity is the maximum number of tokens can be consumed.
   * @param tokensGeneratedInPeriod Tokens generated in period.
   * @param period                  Period that generating specific number of tokens.
   */
  public BucketRateLimiter(long capacity, long tokensGeneratedInPeriod, Duration period) {
    Bandwidth bandwidth =
            Bandwidth.classic(capacity, Refill.greedy(tokensGeneratedInPeriod, period));

    this.bucket = Bucket4j.builder().
            addLimit(bandwidth).
            withSynchronizationStrategy(SynchronizationStrategy.SYNCHRONIZED).
            build();
  }

  @Override
  public synchronized Duration when(T item) {
    DelayGetter delayGetter = new DelayGetter();
    bucket.asAsyncScheduler().consume(1, delayGetter).complete(null);
    return delayGetter.getDelay();
  }

  @Override
  public void forget(T item) {
  }

  @Override
  public int numRequeues(T item) {
    return 0;
  }

  private class DelayGetter extends ScheduledThreadPoolExecutor {
    private Duration delay = Duration.ZERO;

    @Override
    public ScheduledFuture<?> schedule(Runnable command, long delay, TimeUnit unit) {
      this.delay = Duration.ofNanos(unit.toNanos(delay));
      return null;
    }

    public DelayGetter() {
      super(0);
    }

    public Duration getDelay() {
      return delay;
    }
  }
}

```

cizezsy replied:
oops, just realized my implementation is wrong :(

I will replace with your implementation, thanks!


withSynchronizationStrategy(SynchronizationStrategy.SYNCHRONIZED).
 

adohe commented on extended/src/main/java/io/kubernetes/client/extended/workqueue/ratelimiter/BucketRateLimiter.java:
```diff
@@ -0,0 +1,56 @@
+package io.kubernetes.client.extended.workqueue.ratelimiter;
+
+import io.github.bucket4j.*;
```
please not using wild card in Java import statement, more info could be found here: https://stackoverflow.com/questions/147454/why-is-using-a-wild-card-with-a-java-import-statement-bad

cizezsy replied:
OK, will replace with single class imports.

brendandburns commented on extended/src/main/java/io/kubernetes/client/extended/workqueue/ratelimiter/ItemExponentialFailureRateLimiter.java:
```diff
@@ -0,0 +1,51 @@
+    failures.put(item, exp + 1);
+
+    double backOff = baseDelay.toNanos() * Math.pow(2, exp);
+    if (backOff > Long.MAX_VALUE) {
```
Shouldn't this be:
`if (backOff > maxDelay) {` ...

Or alternately

`backoff = Math.min(backoff, maxDelay);`

cizezsy replied:
Yeah, you are right, will change it :)

adohe commented on extended/src/main/java/io/kubernetes/client/extended/workqueue/ratelimiter/ItemFastSlowRateLimiter.java:
```diff
@@ -0,0 +1,48 @@
+  }
+
+  @Override
+  public synchronized Duration when(T item) {
```
same as above, not using synchronized method here.

brendandburns replied:
again `ConcurrentHashMap` and `AtomicInteger` will provide what you need.

cizezsy replied:
Replaced with AtomicLongMap.

cizezsy commented:
Sorry for the delay, I am too busy in last week....

I have changed the implementation of `BucketRateLimiter` and `ItemExponentialFailureRateLimiter` as comment, and added some unit tests.

yue9944882 commented:
LGTM, @brendandburns may want another round of review?

brendandburns commented on extended/src/main/java/io/kubernetes/client/extended/workqueue/ratelimiter/BucketRateLimiter.java:
```diff
@@ -0,0 +1,64 @@
+    @Override
+    public ScheduledFuture<?> schedule(Runnable command, long delay, TimeUnit unit) {
+      this.delay = Duration.ofNanos(unit.toNanos(delay));
+      return null;
```
Should this really return null?

cizezsy replied:
@brendanburns 

The return value will not be used by Bucket4j according to the comment of https://github.com/vladimir-bukhtoyarov/bucket4j/issues/96, so I think it ok to return a null value.
>  ScheduledFuture that returned also not used by Bucket4j, so do not spend efforts to provide correctly result.

brendandburns commented:
A few additional comments.

yue9944882 commented on extended/src/main/java/io/kubernetes/client/extended/workqueue/ratelimiter/BucketRateLimiter.java:
```diff
@@ -0,0 +1,64 @@
+  }
+
+  @Override
+  public synchronized Duration when(T item) {
```
```suggestion
  public Duration when(T item) {
```

no need for the modifier, isn't it?

cizezsy replied:
Yes, will remove it.

yue9944882 commented:
/lgtm
/approve
/hold

holding for lazy consensus with @brendandburns

brendandburns commented:
/hold remove

brendandburns commented:
/hold cancel
